src/ucp/core/ucp_context.h

```diff
@@ -497,25 +497,21 @@ ucp_memory_detect_internal(ucp_context_h context, const void *address,
     }
 
     status = ucs_memtype_cache_lookup(address, length, mem_info);
-    if (status != UCS_ERR_NO_ELEM) {
-        if (ucs_likely(status != UCS_OK)) {
-            ucs_assert(status == UCS_ERR_NO_ELEM);
-            goto out_host_mem;
-        }
-
-        if ((mem_info->type != UCS_MEMORY_TYPE_UNKNOWN) &&
-            ((mem_info->sys_dev != UCS_SYS_DEVICE_ID_UNKNOWN))) {
-            return;
-        }
-
-        /* Fall thru to slow-path memory type and system device detection by UCT
-         * memory domains. In any case, the memory type cache is not expected to
-         * return HOST memory type.
-         */
-        ucs_assert(mem_info->type != UCS_MEMORY_TYPE_HOST);
+    if (ucs_likely(status == UCS_ERR_NO_ELEM)) {
+        goto out_host_mem;
+    }
+
+    ucs_assertv((status == UCS_OK) || (status == UCS_ERR_UNSUPPORTED),
```

Comment:
what is the advantage? it adds another branch (line 507)

Reply:
> what is the advantage? it adds another branch (line 507)

it fixes the problem with memory type detection if MEMTYPE cache wasn't enabled - we should go to memory detection through MDs. without this check - we assumed that memory type is HOST.

Reply:
interesting...
maybe we should set mem_info->type=UNKNOWN in this case?

Reply:
> 

actually it would be weird to set mem_info in case of error return,
so guess better restore an equivalent of the prev logic where we checked if memtype_cache was enabled in the first place. keep it UNSUPPORTED

Reply:
> so guess better restore an equivalent of the prev logic where we checked if memtype_cache was enabled in the first place. keep it UNSUPPORTED

do you mean checking `status == UCS_ERR_UNSUPPORTED` and doing memory detection in this case?

Reply:
yes

Reply:
done